torchvision/tv_tensors/__init__.py

```diff
@@ -8,6 +8,10 @@
 from ._video import Video
 
 
+# TODO: Fix this. We skip this method as it leads to
+# RecursionError: maximum recursion depth exceeded while calling a Python object
+# Keeping it here, leads to graph breaks between multiple functional ops instead of having a single graph
+@torch.compiler.disable
```

Comment:
Just to confirm, the issue only appears for subclasses but not for pure tensors?
Was this creating a graph break for pure tensors? Because if not, it might be best to ignore that altogether for now?

Reply:
Yes, on subclasses when we try to convert result tensor back into its subclass.

> Was this creating a graph break for pure tensors?

This method is not called for pure tensors.

> Because if not, it might be best to ignore that altogether for now?

With this added we can torch compile functional ops with subclasses vs only image processing functional ops on tensors:
```python
from torchvision.transforms.v2.functional import horizontal_flip
from torchvision.tv_tensors import Image, BoundingBoxes

t_img = torch.randint(0, 256, size=(3, 46, 52), dtype=torch.uint8)
img = Image(t_img)
box = BoundingBoxes(torch.randint(0, 256, size=(5, 4), dtype=torch.uint8))

cfn = torch.compile(horizontal_flip)
o1 = cfn(t_img)
o2 = cfn(img)
o3 = cfn(box)
```
vs 
```python
from torchvision.transforms.v2.functional import horizontal_flip
from torchvision.tv_tensors import Image, BoundingBoxes

t_img = torch.randint(0, 256, size=(3, 46, 52), dtype=torch.uint8)
img = Image(t_img)
box = BoundingBoxes(torch.randint(0, 256, size=(5, 4), dtype=torch.uint8))

cfn = torch.compile(horizontal_flip)
o1 = cfn(t_img)  # OK
o2 = cfn(img)    # RecursionError: maximum recursion depth exceeded while calling a Python object
o3 = cfn(box)    # RecursionError: maximum recursion depth exceeded while calling a Python object
```
I think this is valuable. 

The only drawback is that we can't make a single graph for a pipeline of ops:
```python
from torchvision.transforms.v2.functional import horizontal_flip, vertical_flip
from torchvision.tv_tensors import Image

def func(x):
    x = horizontal_flip(x)
    x = vertical_flip(x)
    return x

cfn = torch.compile(func)

x = Image(torch.randint(0, 256, size=(3, 46, 52), dtype=torch.uint8))
out = cfn(x)  # produces 4 graphs
```
vs just a failure.

Hope it is more clear.

Reply:
Thanks for the explanation!

Should we... Merge that fix right now separately from the tests? I agree this valuable to have to avoid failure. But I'm a bit more concerned about adding the test (at least for now) because:

- they add a lot of time (1min for testing tensors only, 4mins if we test all subclasses)
- checking that 1 or 0 graph break happens doesn't add a ton of value over what we already know TBH. It'd be more valuable if we could assert there are 0 graph breaks so that our test ensures there's never a regression (right now, it won't catch a functional that goes from 0 to 1 graph break).
- `explain` has proven to be unreliable in counting graph breaks e.g. in https://github.com/pytorch/vision/issues/8056#issue-1955064062 (although IDK what else we could be using).

So I'm tempted to merge the fix but not merge the tests for now, and keep this PR open to keep track of the progress on the graph breaks. Then when we're in a better spot (e.g. no graph break across all functionals for tensors) maybe we can reconsider? WDYT @vfdev-5 @pmeier ?

Reply:
:+1: on merging the fix. 

I don't really understand why we need to except 0 or 1 graph breaks though. Can't we assert that kernels have 0 graph breaks and functionals 1 (until https://github.com/pytorch/vision/issues/8056#issuecomment-1803581357) and xfail / disable the ones that currently don't pass these checks?

And since `explain` has its own set of issues, maybe we just run the test on kernels with `full_graph=True`? That way we circumvent the `explain` issue. Plus, we don't lose much value in doing so, since our functionals for the most part are just boilerplate the same. Meaning, if the specific kernel works and functionals work in general, the specific functional should work as well.

Reply:
OK, let's add `@torch.compiler.disable` in another PR.

In the tests we can disable checking on subclasses and run checks on tensor images input only to reduce CI time.

> right now, it won't catch a functional that goes from 0 to 1 graph break

Right now if kernel is not empty we always have 1 break. So, if non-empty kernel starts creating 2 graphs we'll see this. What we can do now is to set it to 1 and update kwargs on tests which produce empty kernels and thus 1 empty graph.

> So I'm tempted to merge the fix but not merge the tests for now, and keep this PR open to keep track of the progress on the graph breaks. 

Is it possible to go as merging this PR (in two parts) and see the CI fail once we got rid of dict look-up graph break ?

Reply:
Why would CI fail when we get rid of the dict look-up graph break? If we're asserting that there is 0 or 1 graph break, tests will always be green (and thus not very useful)?

Reply:
We can set expected number of graph breaks to 1 and update kwargs such that we never generate an empty graph, so that it will be broken by dict look-up.